File: os02_process/time_fork_exec_wait.py
import os
import sys
import time

def cur_time():
    return int(time.time() * 1.0e9)

def main():
    n = int(sys.argv[1]) if 1 < len(sys.argv) else 5
    t0 = cur_time()

    for i in range(n):
        pid = os.fork()
        if pid == 0:
            argv = ["./do_nothing"]
            os.execv(argv[0], argv)
            sys.exit()
        else:
            os.waitpid(pid, 0)
  
    t1 = cur_time()
    dt = t1 - t0
    print(f"{dt} nsec to fork-exec-wait {n} processes ({dt / n} nsec/proc)")

main()

File: os02_process/time_fork_exec_wait.c
#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <sys/types.h>
#include <sys/wait.h>

long cur_time() {
  struct timespec ts[1];
  clock_gettime(CLOCK_REALTIME, ts);
  return ts->tv_sec * 1000000000L + ts->tv_nsec;
}

int main(int argc, char ** argv) {
    int n = (argc > 1 ? atoi(argv[1]) : 5);
    long t0 = cur_time();

    for (int i=0;i<n;i++){
        pid_t pid = fork();
        if (pid == -1) {
            err(1, "fork");
        } else if (pid == 0) {
            char* const argv[] = {"./do_nothing"};
            execv(argv[0], argv);
            exit(0);
        } else {
            int ws;
            pid_t cid = waitpid(pid, &ws, 0);
            if (cid == -1) err(1, "waitpid");
        }
    }

    long t1 = cur_time();
    long dt = t1 - t0;
    printf("%ld nsec to fork-exec-wait %d processes (%ld nsec/proc)\n", dt, n, dt / n);
    return 0;
}